File: src/entities/transfer-process.ts
import { DataAddress } from "./data-address";
import { JsonLdId, JsonLdObject } from "./jsonld";

export class TransferProcess extends JsonLdId {

  get correlationId(): string {
    return this.mandatoryValue('edc', 'correlationId');
  }

  get type(): string {
    return this.mandatoryValue('edc', 'type');
  }

  get state(): string {
    return this.mandatoryValue('edc', 'state');
  }

  get createdAt(): number {
    return this.mandatoryValue('edc', 'createdAt');
  }

  get assetId(): string {
    return this.mandatoryValue('edc', 'assetId');
  }

  get connectorId(): string {
    return this.mandatoryValue('edc', 'connectorId');
  }

  get contractId(): string {
    return this.mandatoryValue('edc', 'contractId');
  }

  get errorDetail(): string | undefined {
    return this.optionalValue('edc', 'errorDetail');
  }

  get dataDestination(): JsonLdObject {
    return this.nested('edc', 'dataDestination');
  }

  get privateProperties(): JsonLdObject {
    return this.nested('edc', 'privateProperties');
  }

}

export class TransferProcessState extends JsonLdObject {
  get state(): string {
    return this.mandatoryValue('edc', 'state');
  }
}

export enum TransferProcessStates {
  INITIAL = "INITIAL",
  PROVISIONING = "PROVISIONING",
  PROVISIONING_REQUESTED = "PROVISIONING_REQUESTED",
  PROVISIONED = "PROVISIONED",
  REQUESTING = "REQUESTING",
  REQUESTED = "REQUESTED",
  STARTING = "STARTING",
  STARTED = "STARTED",
  SUSPENDING = "SUSPENDING",
  SUSPENDED = "SUSPENDED",
  COMPLETING = "COMPLETING",
  COMPLETED = "COMPLETED",
  TERMINATING = "TERMINATING",
  TERMINATED = "TERMINATED",
  DEPROVISIONING = "DEPROVISIONING",
  DEPROVISIONING_REQUESTED = "DEPROVISIONING_REQUESTED",
  DEPROVISIONED = "DEPROVISIONED",
}

export interface TransferProcessInput {
  assetId: string;
  connectorAddress: string;
  connectorId: string;
  contractId: string;
  dataDestination: DataAddress;
  id?: string;
  properties?: Record<string, string>;
}

export interface EndpointDataReference {
  id: string;
  endpoint: string;
  contractId: string;
  authKey: string;
  authCode: string;
  properties?: Record<string, string>;
}

File: rust/src/entities/transfer_process.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use super::data_address::DataAddress;

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferProcess {
    pub created_at: u64,
    pub updated_at: u64,
    pub id: String,
    #[serde(rename = "type")]
    pub kind: String, // TODO(@fdionisi): specialise
    pub state: String, // TODO(@fdionisi): specialise
    pub state_timestamp: u64,
    pub error_detail: Option<String>,
    pub data_request: TransferProcessDataRequest,
    data_destination: TransferProcessDataDestination,
}

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferProcessDataRequest {
    pub id: String,
    pub asset_id: String,
    pub contract_id: String,
    pub connector_id: String,
}
#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferProcessDataDestination {
    pub properties: HashMap<String, String>,
}

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferProcessInput {
    pub asset_id: String,
    pub connector_address: String,
    pub connector_id: String,
    pub contract_id: String,
    pub data_destination: DataAddress,
    pub protocol: Option<String>,
    pub transfer_type: Option<TransferType>,
    pub id: Option<String>,
    pub properties: Option<HashMap<String, String>>,
    pub managed_resources: bool,
}

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferType {
    pub content_type: Option<String>,
    pub is_finite: Option<bool>,
}

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferProcessResponse {
    pub id: String,
    pub endpoint: String,
    pub contract_id: String,
    pub auth_key: String,
    pub auth_code: String,
    pub properties: Option<HashMap<String, String>>,
}
